### PR TITLE
LF: Drop Builtin Exceptions completly

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 1]
```diff
@@ -176,15 +176,14 @@ private[speedy] sealed abstract class SBuiltin(val arity: Int) {
         )
     }
 
-  final protected def getSException(args: util.ArrayList[SValue], i: Int): SException =
+  final protected def getSException(args: util.ArrayList[SValue], i: Int): SAnyException =
```
Comment (sofiafaro-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:179]: Might be worth renaming this function `getSAnyException`

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala [hunk 2]
```diff
@@ -207,13 +203,29 @@ object SValue {
...
+      val array = new util.ArrayList[SValue](1)
+      array.add(
+        SText(s"ArithmeticError while evaluating ($builtinName ${args.iterator.mkString(" ")}).")
+      )
+      SAnyException(typ, SRecord(tyCon, fields, array))
+    }
+    // Assumes excep is properly typed
+    def unapply(excep: SAnyException): Option[SValue] =
+      excep match {
+        case SAnyException(`typ`, SRecord(_, _, args)) =>
```
Comment (sofiafaro-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala:224]: Scala Q: Is this matching against the `typ` value?

File: compiler/damlc/tests/daml-test-files/ExceptionSemantics.daml [hunk 3]
```diff
@@ -3,7 +3,7 @@
 
 -- @SINCE-LF-FEATURE DAML_EXCEPTIONS
 -- @ERROR range=130:1-130:23; Unhandled exception: ExceptionSemantics:E
--- @ERROR range=145:1-145:25; Unhandled exception: ArithmeticError while evaluating (DIV_INT64 1 0).
+-- @ERROR range=145:1-145:25; ArithmeticError while evaluating (DIV_INT64 1 0).
```
Comment (cocreature) [inline compiler/damlc/tests/daml-test-files/ExceptionSemantics.daml:6]: Can we keep the full error here to make sure this gets propagated as an `Unhandled exception`?

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala (hunk 2 above)
Comment (cocreature) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala:224]: Yes

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala [hunk 4]
```diff
@@ -207,13 +203,29 @@ object SValue {
...
-  final case class SArithmeticError(
-      builtinName: String,
-      args: ImmArray[String],
-  ) extends SException {
-    def message = s"ArithmeticError while evaluating ($builtinName ${args.iterator.mkString(" ")})."
+  final case class SAnyException(ty: Type, value: SValue) extends SValue
+
+  object SArithmeticError {
+    val tyCon: Ref.TypeConName = Ref.Identifier.assertFromString(
+      "f1cf1ff41057ce327248684089b106d0a1f27c2f092d30f663c919addf173981:DA.Exception.ArithmeticError:ArithmeticError"
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala:210]: maybe add a comment that this should be kept in sync with the generator or the tests.

**Review by cocreature — APPROVED** [review]
Nice! The one issue in this approach that I could come up with after thinking about it for a while is that if we start exposing unhandled exceptions on the ledger API we will produce values for which the ledger might not have a Dalf. I don’t expect that to be that big of an issue in practice but it is a bit annoying.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala (hunk 2 above)
Comment (sofiafaro-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala:224]: Thanks for the explanation :-)

**Review by cocreature — APPROVED** [review]
Thanks!

**Pull request merged**